File: packages/python/src/agent_wallet/core/providers/local.py
"""Local wallet provider implementation."""

from __future__ import annotations

from pathlib import Path

from agent_wallet.core.base import BaseWallet, WalletType
from agent_wallet.core.errors import WalletNotFoundError
from agent_wallet.core.providers.base import WalletProvider
from agent_wallet.local.config import (
    WalletConfig,
    WalletInfo,
    load_config,
    save_config,
)
from agent_wallet.local.kv_store import SecureKVStore


class LocalWalletProvider(WalletProvider):
    """Local mode: decrypt all wallets at init from secrets/ directory."""

    def __init__(self, secrets_dir: str | Path, password: str) -> None:
        self._secrets_dir = secrets_dir
        kv_store = SecureKVStore(secrets_dir, password)
        kv_store.verify_password()
        self._config = load_config(secrets_dir)
        self._wallets: dict[str, BaseWallet] = {}
        for wid, conf in self._config.wallets.items():
            self._wallets[wid] = _create_wallet(conf, kv_store)

    async def list_wallets(self) -> list[WalletInfo]:
        return [
            WalletInfo(id=wid, type=conf.type)
            for wid, conf in self._config.wallets.items()
        ]

    async def get_wallet(self, wallet_id: str) -> BaseWallet:
        if wallet_id not in self._wallets:
            raise WalletNotFoundError(f"Wallet '{wallet_id}' not found")
        return self._wallets[wallet_id]

    def get_active_id(self) -> str | None:
        return self._config.active_wallet

    async def get_active_wallet(self) -> BaseWallet:
        active_id = self.get_active_id()
        if not active_id:
            raise WalletNotFoundError(
                "No active wallet set. Use 'agent-wallet use <id>' to set one."
            )
        return await self.get_wallet(active_id)

    async def get_active(self) -> BaseWallet:
        """Backward-compatible alias for older SDK callers."""
        return await self.get_active_wallet()

    def set_active(self, wallet_id: str) -> None:
        if wallet_id not in self._wallets:
            raise WalletNotFoundError(f"Wallet '{wallet_id}' not found")
        self._config.active_wallet = wallet_id
        save_config(self._secrets_dir, self._config)


def _create_wallet(conf: WalletConfig, kv_store: SecureKVStore) -> BaseWallet:
    """Instantiate the correct wallet adapter based on config type."""
    match conf.type:
        case WalletType.EVM_LOCAL:
            from agent_wallet.core.adapters.evm import EvmWallet

            assert conf.identity_file is not None
            private_key = kv_store.load_private_key(conf.identity_file)
            return EvmWallet(private_key=private_key)

        case WalletType.TRON_LOCAL:
            from agent_wallet.core.adapters.tron import TronWallet

            assert conf.identity_file is not None
            private_key = kv_store.load_private_key(conf.identity_file)
            return TronWallet(private_key=private_key)

        case _:
            raise ValueError(f"Unknown or unimplemented wallet type: {conf.type}")

File: packages/typescript/src/core/providers/local.ts
import type { BaseWallet } from '../base.js'
import { EvmWallet } from '../adapters/evm.js'
import { TronWallet } from '../adapters/tron.js'
import { WalletNotFoundError } from '../errors.js'
import { loadConfig, saveConfig, type WalletConfig, type WalletInfo, type WalletsTopology } from '../../local/config.js'
import { SecureKVStore } from '../../local/kv-store.js'
import { WalletProvider } from './base.js'

export class LocalWalletProvider extends WalletProvider {
  private readonly config: WalletsTopology
  private readonly wallets: Map<string, BaseWallet> = new Map()

  constructor(
    private readonly secretsDir: string,
    password: string,
  ) {
    super()
    const kvStore = new SecureKVStore(secretsDir, password)
    kvStore.verifyPassword()
    this.config = loadConfig(secretsDir)
    for (const [wid, conf] of Object.entries(this.config.wallets)) {
      this.wallets.set(wid, createWallet(conf, kvStore))
    }
  }

  async listWallets(): Promise<WalletInfo[]> {
    return Object.entries(this.config.wallets).map(([wid, conf]) => ({
      id: wid,
      type: conf.type,
    }))
  }

  async getWallet(walletId: string): Promise<BaseWallet> {
    const wallet = this.wallets.get(walletId)
    if (!wallet) {
      throw new WalletNotFoundError(`Wallet '${walletId}' not found`)
    }
    return wallet
  }

  getActiveId(): string | null {
    return this.config.active_wallet ?? null
  }

  async getActiveWallet(): Promise<BaseWallet> {
    const activeId = this.getActiveId()
    if (!activeId) {
      throw new WalletNotFoundError("No active wallet set. Use 'agent-wallet use <id>' to set one.")
    }
    return this.getWallet(activeId)
  }

  async getActive(): Promise<BaseWallet> {
    return this.getActiveWallet()
  }

  setActive(walletId: string): void {
    if (!this.wallets.has(walletId)) {
      throw new WalletNotFoundError(`Wallet '${walletId}' not found`)
    }
    this.config.active_wallet = walletId
    saveConfig(this.secretsDir, this.config)
  }
}

function createWallet(conf: WalletConfig, kvStore: SecureKVStore): BaseWallet {
  switch (conf.type) {
    case 'evm_local': {
      const privateKey = kvStore.loadPrivateKey(conf.identity_file!)
      return new EvmWallet(privateKey)
    }
    case 'tron_local': {
      const privateKey = kvStore.loadPrivateKey(conf.identity_file!)
      return new TronWallet(privateKey)
    }
    default:
      throw new Error(`Unknown wallet type: ${conf.type}`)
  }
}
